### PR TITLE
batch sys: at: poll exit, poll PID is alive or not

### LINKED ISSUE #1661: atq on AIX is useless to us
The joys of AIX :grimacing:.   In addition to https://github.com/cylc/cylc/issues/1659#issuecomment-153193433, `atq` on AIX does not list running jobs, only queued jobs yet to start running are reported. Cylc job poll for `at` jobs on AIX therefore reports job failed even if the job is running.  It's not particularly likely that anyone is submitting jobs to `at` on AIX, but for completeness we should probably have some way of telling cylc that such jobs cannot be polled.

### PR DESCRIPTION
**Pull request by matthewrmshin**
When polling "at" jobs, if a job appears to have exited "atq", check PID
of job to see if it is still alive. In some systems, the job may not be
on "atq" while it is still running as a normal Unix process.

Close #1661.

### CODE REVIEW

File: lib/cylc/batch_sys_manager.py [hunk 1]
```diff
@@ -592,44 +600,59 @@ def _jobs_poll_status_files(self, job_log_root, job_log_dir):
...
-                if head in all_job_ids:
-                    job_ids.append(head)
+        exp_job_ids = [ctx.batch_sys_job_id for ctx in my_ctx_list]
+        bad_job_ids = list(exp_job_ids)
+        exp_pids = []
+        bad_pids = []
+        items = [[self.get_inst(batch_sys_name), exp_job_ids, bad_job_ids]]
+        if getattr(items[0][0], "CAN_POLL_PROC_GROUP", False):
+            exp_pids = [ctx.pid for ctx in my_ctx_list if ctx.pid is not None]
+            bad_pids = bad_pids.extend(exp_pids)
```
Comment (hjoliver) [inline lib/cylc/batch_sys_manager.py]: This is wrong (no return value)

File: lib/cylc/batch_sys_manager.py [hunk 2]
```diff
@@ -72,9 +72,13 @@
...
 
 batch_sys.CAN_KILL_PROC_GROUP
-    * A boolean to indicate whether it is possible to kill the job by sending
+    * A boolean to indicate whether it is possible to kill a job by sending
       a signal to its Unix process group.
 
+batch_sys.CAN_POLL_PROC_GROUP
+    * A boolean to indicate whether it is necessary to poll a job by its PID
+      as well as the job ID.
+
```
Comment (hjoliver) [inline lib/cylc/batch_sys_manager.py]: Should this be `SHOULD_POLL_PROCESS_GROUP`?  (you "can" poll any job by PID, but for at we need to do it)

File: lib/cylc/batch_sys_manager.py [hunk 3]
```diff
@@ -592,44 +600,59 @@ def _jobs_poll_status_files(self, job_log_root, job_log_dir):
...
+        bad_job_ids = list(exp_job_ids)
+        exp_pids = []
+        bad_pids = []
+        items = [[self.get_inst(batch_sys_name), exp_job_ids, bad_job_ids]]
+        if getattr(items[0][0], "CAN_POLL_PROC_GROUP", False):
+            exp_pids = [ctx.pid for ctx in my_ctx_list if ctx.pid is not None]
+            bad_pids = bad_pids.extend(exp_pids)
+            items.append([self.get_inst("background"), exp_pids, bad_pids])
+        for batch_sys, exp_ids, bad_ids in items:
+            batch_sys = self.get_inst(batch_sys_name)
```
Comment (hjoliver) [inline lib/cylc/batch_sys_manager.py]: This line needs to be deleted - it reverts `batch_sys` back to the original ("at").

**Comment by hjoliver** [comment]
This seems to work correctly on AIX with the two fixes above.

**Comment by matthewrmshin** [comment]
Suggested logic fixes implemented.

**Comment by matthewrmshin** [comment]
@hjoliver please review 1. I am happy to write a test for this (although it will probably always returns positive at this end).

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by hjoliver** [comment]
Review 1 - good, works now.

**Comment by matthewrmshin** [comment]
@arjclark please sanity check.

**Comment by arjclark** [comment]
Looks sane enough to me. Test battery passing in my environment.

**Pull request merged**